controllers/actions.go

```diff
@@ -0,0 +1,160 @@
+	utils "github.com/openshift-kni/cluster-group-upgrades-operator/controllers/utils"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	clusterv1 "open-cluster-management.io/api/cluster/v1"
+)
+
+// takeActionsBeforeEnable takes the required actions before starting upgrade
+// returns: error/nil
+func (r *ClusterGroupUpgradeReconciler) takeActionsBeforeEnable(
+	ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) error {
+
+	actionsBeforeEnable := clusterGroupUpgrade.Spec.Actions.BeforeEnable
+	// Add/delete cluster labels
+	if actionsBeforeEnable.AddClusterLabels != nil || actionsBeforeEnable.DeleteClusterLabels != nil {
+		clusters, err := r.getAllClustersForUpgrade(ctx, clusterGroupUpgrade)
+		if err != nil {
+			return err
+		}
+		labels := map[string]map[string]string{
+			"add":    actionsBeforeEnable.AddClusterLabels,
+			"delete": actionsBeforeEnable.DeleteClusterLabels,
+		}
+
+		if err := r.manageClusterLabels(ctx, clusters, labels); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// takeActionsAfterCompletion takes the required actions after upgrade is completed
+// returns: error/nil
+func (r *ClusterGroupUpgradeReconciler) takeActionsAfterCompletion(
+	ctx context.Context, clusterGroupUpgrade *ranv1alpha1.ClusterGroupUpgrade) error {
+
+	actionsAfterCompletion := clusterGroupUpgrade.Spec.Actions.AfterCompletion
+	// Add/delete cluster labels
+	if actionsAfterCompletion.AddClusterLabels != nil || actionsAfterCompletion.DeleteClusterLabels != nil {
+		clusters, err := r.getAllClustersForUpgrade(ctx, clusterGroupUpgrade)
+		if err != nil {
+			return err
+		}
+		labels := map[string]map[string]string{
+			"add":    actionsAfterCompletion.AddClusterLabels,
+			"delete": actionsAfterCompletion.DeleteClusterLabels,
+		}
+
+		if err := r.manageClusterLabels(ctx, clusters, labels); err != nil {
+			return err
+		}
+	}
+
+	// Cleanup resources
+	if *actionsAfterCompletion.DeleteObjects {
+		err := r.deleteResources(ctx, clusterGroupUpgrade)
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// manageClusterLabels adds/deletes the cluster labels for selected clusters
+func (r *ClusterGroupUpgradeReconciler) manageClusterLabels(
+	ctx context.Context, clusters []string, labels map[string]map[string]string) error {
+
+	for _, c := range clusters {
+		managedCluster := &clusterv1.ManagedCluster{}
+		if err := r.Get(ctx, types.NamespacedName{Name: c}, managedCluster); err != nil {
+			if errors.IsNotFound(err) {
+				continue
+			}
+			return err
+		}
+		if err := r.addClusterLabels(ctx, managedCluster, labels["add"]); err != nil {
+			return fmt.Errorf("Fail to add labels for cluster %s: %v", managedCluster.Name, err)
+		}
+		if err := r.deleteClusterLabels(ctx, managedCluster, labels["delete"]); err != nil {
+			return fmt.Errorf("Fail to delete labels for cluster %s: %v", managedCluster.Name, err)
+		}
```

Comment:
Does this generate an error if the label does not exist?  If this fails because the cluster does not have the specified label then we should not error out. The intent of the "deleteLables" is that the cluster should not have the label, which is the case if it does not exist.

Reply:
No error for the non-existence label. Delete is a no-op if label doesn't exist. I added two more unittests to cover this.